In [1]:
pip install tensorflow

     |████████████████████████████████| 320.4 MB 25 kB/s  eta 0:00:01     |███████████████▋                | 156.8 MB 3.0 MB/s eta 0:00:54
     |████████████████████████████████| 127 kB 58.5 MB/s eta 0:00:01
     |████████████████████████████████| 20.1 MB 50.5 MB/s eta 0:00:01    |████▋                           | 2.9 MB 50.5 MB/s eta 0:00:01
     |████████████████████████████████| 459 kB 60.0 MB/s eta 0:00:01
     |████████████████████████████████| 6.8 MB 50.1 MB/s eta 0:00:01
     |████████████████████████████████| 42 kB 2.0 MB/s  eta 0:00:01
     |████████████████████████████████| 65 kB 5.4 MB/s  eta 0:00:01
     |████████████████████████████████| 1.3 MB 50.9 MB/s eta 0:00:01
     |████████████████████████████████| 3.8 MB 52.2 MB/s eta 0:00:01
     |████████████████████████████████| 2.9 MB 65.6 MB/s eta 0:00:01
     |████████████████████████████████| 57 kB 6.1 MB/s  eta 0:00:01
     |████████████████████████████████| 779 kB 41.6 MB/s eta 0:00:01
     |███████████████████████████████

In [2]:
pip install keras

Note: you may need to restart the kernel to use updated packages.


Load the data in pandas data frame.

In [1]:
import pandas as pd

url1="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_554093d294594a7e9051e3da088590a3.csv"
df1=pd.read_csv(url1,header=None)
df1.head()

0      1      2      3     4       5      6      7       8      9    ...  \
0    0   87.0  975.0 -287.0  11.0  1001.0  163.0   95.0   975.0  152.0  ...   
1   33  124.0  978.0 -389.0  -7.0  1014.0  199.0  124.0   968.0  123.0  ...   
2   67  102.0  996.0 -440.0 -49.0  1024.0  193.0  127.0  1001.0  113.0  ...   
3  100   59.0  861.0 -384.0  -9.0  1023.0  202.0  110.0  1007.0  106.0  ...   
4  133  119.0  946.0 -426.0 -22.0  1026.0  188.0   98.0  1001.0   92.0  ...   

      240     241     242  243  244  245  246  247  248  249  
0  5789.0  2907.0  1447.0    0    0    0    0    0    0    0  
1  5789.0  2908.0  1443.0    0    0    0    0    0    0    0  
2  5789.0  2910.0  1440.0    0    0    0    0    0    0    0  
3  5789.0  2912.0  1440.0    0    0    0    0    0    0    0  
4  5791.0  2915.0  1442.0    0    0    0    0    0    0    0  

[5 rows x 250 columns]

In [2]:
df1.shape

(51116, 250)

Separate the predictor variables (X) and target variables (y).

In [5]:
X=df1.iloc[:,0:243]
X.head()

0      1      2      3     4       5      6      7       8      9    ...  \
0    0   87.0  975.0 -287.0  11.0  1001.0  163.0   95.0   975.0  152.0  ...   
1   33  124.0  978.0 -389.0  -7.0  1014.0  199.0  124.0   968.0  123.0  ...   
2   67  102.0  996.0 -440.0 -49.0  1024.0  193.0  127.0  1001.0  113.0  ...   
3  100   59.0  861.0 -384.0  -9.0  1023.0  202.0  110.0  1007.0  106.0  ...   
4  133  119.0  946.0 -426.0 -22.0  1026.0  188.0   98.0  1001.0   92.0  ...   

      233     234     235     236     237     238     239     240     241  \
0  1265.0  5858.0  2707.0  1134.0  6294.0  2663.0  1533.0  5789.0  2907.0   
1  1267.0  5905.0  2708.0  1155.0  6294.0  2664.0  1532.0  5789.0  2908.0   
2  1268.0  5948.0  2710.0  1175.0  6295.0  2664.0  1531.0  5789.0  2910.0   
3  1266.0  5986.0  2711.0  1192.0  6296.0  2664.0  1530.0  5789.0  2912.0   
4  1263.0  6018.0  2712.0  1205.0  6299.0  2664.0  1529.0  5791.0  2915.0   

      242  
0  1447.0  
1  1443.0  
2  1440.0  
3  1440.0  
4  1442.0  

[5 rows x 243 columns]

In [6]:
y=df1.iloc[:,243:250]
y.head()

243  244  245  246  247  248  249
0    0    0    0    0    0    0    0
1    0    0    0    0    0    0    0
2    0    0    0    0    0    0    0
3    0    0    0    0    0    0    0
4    0    0    0    0    0    0    0

Check if the dataset has empty cells.

In [7]:
X.isna().sum()

0        0
1      911
2      911
3      911
4      466
      ... 
238    409
239    409
240    409
241    409
242    409
Length: 243, dtype: int64

It looks like there are a lot of empty cells in the predictor variables.

In [8]:
y.isna().sum()

243    0
244    0
245    0
246    0
247    0
248    0
249    0
dtype: int64

There is no empty cells in the target variables.

Calculate the percentage of empty cells in each column in the predictor variables.

In [9]:
p_m =pd.DataFrame(X.isnull().sum() * 100 / len(X)).rename(columns={0:"Percentage"})
p_m=p_m.sort_values(by="Percentage",ascending=False)
p_m

Percentage
36    99.998044
35    99.998044
34    99.998044
219   99.998044
220   99.998044
..          ...
203    0.000000
202    0.000000
201    0.000000
200    0.000000
0      0.000000

[243 rows x 1 columns]

It appears that some columns have as much as 99.99% of empty cells. These columns will be discarded. We will discard columns with more than 20% of missing data.

In [10]:
p_m=p_m[p_m["Percentage"]<=20]
p_m

Percentage
20     4.067220
21     4.067220
19     4.067220
15     2.369121
13     2.369121
..          ...
203    0.000000
202    0.000000
201    0.000000
200    0.000000
0      0.000000

[237 rows x 1 columns]

After discarding columns with more than 20% of missing values, we have 237 columns in the predictor variables X.

Select those columns with at most 20% missing values.

In [11]:
X=X.iloc[:,p_m.index.tolist()]
X.head()

20    21     19     15     13      14     143   139    140    141  ...  \
0  973.0 -41.0  209.0   99.0  249.0  1008.0  411.0 -72.0  682.0 -778.0  ...   
1  956.0 -93.0  245.0   97.0  272.0   963.0  409.0 -72.0  684.0 -770.0  ...   
2  929.0 -69.0  277.0  116.0  213.0  1017.0  410.0 -72.0  684.0 -770.0  ...   
3  958.0 -76.0  245.0  120.0  234.0  1028.0  411.0 -69.0  673.0 -773.0  ...   
4  968.0 -20.0  231.0  189.0  191.0   991.0  411.0 -62.0  684.0 -787.0  ...   

   196  197  206  205  204  203  202  201  200  0    
0    0    0    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    0    0   33  
2    0    0    0    0    0    0    0    0    0   67  
3    0    0    0    0    0    0    0    0    0  100  
4    0    0    0    0    0    0    0    0    0  133  

[5 rows x 237 columns]

Fill those empty cells with the median of the respective column.

In [12]:
X=X.fillna(X.median())

Verified that there is no more missing values in the predictor variables X.

In [13]:
null=pd.DataFrame(X.isna().any()).rename(columns={0:"Null"})
null[null["Null"]==True]

Empty DataFrame
Columns: [Null]
Index: []

All empty cells have been filled.

For this task, we will build a model to predict the locomotion: Stand, Walk, Sit or Lie. That is the first column in the target variables y.

In [14]:
y1=pd.DataFrame(y.iloc[:,0]).rename(columns={243:"locomotion"})
y1

locomotion
0               0
1               0
2               0
3               0
4               0
...           ...
51111           0
51112           0
51113           0
51114           0
51115           0

[51116 rows x 1 columns]

In [15]:
import numpy as np
np.unique(y1)

array([0, 1, 2, 4, 5])

Feature selection.

In [17]:
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=mutual_info_classif, k="all")
fit = test.fit(X, np.asarray(y1).reshape(y1.shape[0],))
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

scores
236  1.331750
134  0.746486
221  0.646843
210  0.619235
196  0.616472
..        ...
226  0.000000
225  0.000000
223  0.000000
26   0.000000
187  0.000000

[237 rows x 1 columns]

In [18]:
scores=scores[scores["scores"]>0]
scores

scores
236  1.331750
134  0.746486
221  0.646843
210  0.619235
196  0.616472
..        ...
229  0.001498
227  0.001041
235  0.000743
234  0.000650
232  0.000375

[230 rows x 1 columns]

In [19]:
X=X.iloc[:,scores.index]
X

0       240    102    133   118    117    82     49     48   \
0            0  5789.0  119.0  175.0  83.0  -76.0  354.0  430.0  457.0   
1           33  5789.0  124.0  175.0  83.0 -164.0  352.0  432.0  457.0   
2           67  5789.0  126.0  175.0  84.0 -156.0  350.0  434.0  456.0   
3          100  5789.0  128.0  175.0  84.0 -160.0  348.0  436.0  456.0   
4          133  5791.0  129.0  175.0  83.0 -166.0  345.0  437.0  456.0   
...        ...     ...    ...    ...   ...    ...    ...    ...    ...   
51111  1703683  5700.0   10.0    8.0 -30.0  -62.0  420.0  145.0   62.0   
51112  1703716  5700.0   10.0    8.0 -30.0  -62.0  420.0  145.0   62.0   
51113  1703750  5700.0   10.0    8.0 -30.0  -62.0  420.0  145.0   62.0   
51114  1703783  5700.0   10.0    8.0 -30.0  -62.0  420.0  145.0   62.0   
51115  1703816  5700.0   10.0    8.0 -30.0  -62.0  420.0  145.0   62.0   

          237  ...    157     227  197  202   223  206  196  200  201  203  
0      6294.0  ...  417.0 -1080.0    0    0 -86.0    0    0    0    0    0  
1      6294.0  ...  416.0 -1070.0    0    0 -86.0    0    0    0    0    0  
2      6295.0  ...  415.0 -1070.0    0    0 -86.0    0    0    0    0    0  
3      6296.0  ...  417.0 -1077.0    0    0 -86.0    0    0    0    0    0  
4      6299.0  ...  416.0 -1076.0    0    0 -86.0    0    0    0    0    0  
...       ...  ...    ...     ...  ...  ...   ...  ...  ...  ...  ...  ...  
51111  6051.0  ...  417.0 -1070.0    0    0 -86.0    0    0    0    0    0  
51112  6051.0  ...  417.0 -1070.0    0    0 -86.0    0    0    0    0    0  
51113  6051.0  ...  417.0 -1070.0    0    0 -86.0    0    0    0    0    0  
51114  6051.0  ...  417.0 -1070.0    0    0 -86.0    0    0    0    0    0  
51115  6051.0  ...  417.0 -1070.0    0    0 -86.0    0    0    0    0    0  

[51116 rows x 230 columns]

Check for class imbalance.

In [20]:
locomotion=pd.concat([X,y1],axis=1)
(locomotion.groupby("locomotion").count().T)/len(locomotion)

locomotion         0         1         2         4         5
0           0.266238  0.437828  0.128003  0.145864  0.022067
240         0.266238  0.437828  0.128003  0.145864  0.022067
102         0.266238  0.437828  0.128003  0.145864  0.022067
133         0.266238  0.437828  0.128003  0.145864  0.022067
118         0.266238  0.437828  0.128003  0.145864  0.022067
..               ...       ...       ...       ...       ...
206         0.266238  0.437828  0.128003  0.145864  0.022067
196         0.266238  0.437828  0.128003  0.145864  0.022067
200         0.266238  0.437828  0.128003  0.145864  0.022067
201         0.266238  0.437828  0.128003  0.145864  0.022067
203         0.266238  0.437828  0.128003  0.145864  0.022067

[230 rows x 5 columns]

The classes that represent the motions are not evenly distributed. Almost 44% of the data belongs to Class 1, while only slightly more than 2% belongs to Class 5. We are going to oversample the minority classes (Classes 2,4,5): duplicate Classes 2 and 4 by 2 times and Class 5 by 10 times.

To ensure that test data consists of strictly data that the model has not seen during the training process, we will split the data between oversampling. In other words, we will only oversample the minority classes in the train data.

In [56]:
locomotion_train=locomotion.sample(frac=0.9)
locomotion_train.shape

(46004, 231)

In [57]:
L_0=locomotion_train[locomotion_train["locomotion"]==0]
L_1=locomotion_train[locomotion_train["locomotion"]==1]
L_2=locomotion_train[locomotion_train["locomotion"]==2]
L_4=locomotion_train[locomotion_train["locomotion"]==4]


L_5=locomotion_train[locomotion_train["locomotion"]==5]
L_5=pd.concat([L_5]*5,axis=0)

In [58]:
locomotion_train=pd.concat([L_0,L_1,L_2,L_4,L_5],axis=0)
locomotion_train.shape

(50048, 231)

In [59]:
(locomotion_train.groupby("locomotion").count().T)/len(locomotion_train)

locomotion         0         1         2         4         5
0           0.244385  0.402274  0.117827  0.134511  0.101003
240         0.244385  0.402274  0.117827  0.134511  0.101003
102         0.244385  0.402274  0.117827  0.134511  0.101003
133         0.244385  0.402274  0.117827  0.134511  0.101003
118         0.244385  0.402274  0.117827  0.134511  0.101003
..               ...       ...       ...       ...       ...
206         0.244385  0.402274  0.117827  0.134511  0.101003
196         0.244385  0.402274  0.117827  0.134511  0.101003
200         0.244385  0.402274  0.117827  0.134511  0.101003
201         0.244385  0.402274  0.117827  0.134511  0.101003
203         0.244385  0.402274  0.117827  0.134511  0.101003

[230 rows x 5 columns]

In [60]:
locomotion_test=locomotion.drop(locomotion.index[locomotion_train.index.tolist()])
locomotion_test.shape

(5112, 231)

In [61]:
(locomotion_test.groupby("locomotion").count().T)/len(locomotion_test)

locomotion         0         1         2         4         5
0           0.269562  0.439554  0.126369  0.141628  0.022887
240         0.269562  0.439554  0.126369  0.141628  0.022887
102         0.269562  0.439554  0.126369  0.141628  0.022887
133         0.269562  0.439554  0.126369  0.141628  0.022887
118         0.269562  0.439554  0.126369  0.141628  0.022887
..               ...       ...       ...       ...       ...
206         0.269562  0.439554  0.126369  0.141628  0.022887
196         0.269562  0.439554  0.126369  0.141628  0.022887
200         0.269562  0.439554  0.126369  0.141628  0.022887
201         0.269562  0.439554  0.126369  0.141628  0.022887
203         0.269562  0.439554  0.126369  0.141628  0.022887

[230 rows x 5 columns]

Because the dataset was split first and then only the train data was oversampled, although the proportions of the classes change in the train data, the proportions of classes in the test data remain the same.

In [62]:
X_train=locomotion_train.drop(["locomotion"],axis=1)
y_train=locomotion_train[["locomotion"]]

In [63]:
X_test=locomotion_test.drop(["locomotion"],axis=1)
y_test=locomotion_test[["locomotion"]]

In [64]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y_train=pd.DataFrame(enc.fit_transform(y_train).toarray())
y_train

0    1    2    3    4
0      1.0  0.0  0.0  0.0  0.0
1      1.0  0.0  0.0  0.0  0.0
2      1.0  0.0  0.0  0.0  0.0
3      1.0  0.0  0.0  0.0  0.0
4      1.0  0.0  0.0  0.0  0.0
...    ...  ...  ...  ...  ...
50043  0.0  0.0  0.0  0.0  1.0
50044  0.0  0.0  0.0  0.0  1.0
50045  0.0  0.0  0.0  0.0  1.0
50046  0.0  0.0  0.0  0.0  1.0
50047  0.0  0.0  0.0  0.0  1.0

[50048 rows x 5 columns]

In [65]:
y_test=pd.DataFrame(enc.fit_transform(y_test).toarray())
y_test

0    1    2    3    4
0     1.0  0.0  0.0  0.0  0.0
1     1.0  0.0  0.0  0.0  0.0
2     1.0  0.0  0.0  0.0  0.0
3     1.0  0.0  0.0  0.0  0.0
4     1.0  0.0  0.0  0.0  0.0
...   ...  ...  ...  ...  ...
5107  1.0  0.0  0.0  0.0  0.0
5108  1.0  0.0  0.0  0.0  0.0
5109  1.0  0.0  0.0  0.0  0.0
5110  1.0  0.0  0.0  0.0  0.0
5111  1.0  0.0  0.0  0.0  0.0

[5112 rows x 5 columns]

In [66]:
from sklearn import preprocessing
scaler=preprocessing.StandardScaler()
X_train=scaler.fit_transform(X_train)
X_train=pd.DataFrame(X_train)
X_train.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


0         1         2         3         4         5         6    \
0 -1.599022  0.017545  0.911041  1.565083  0.052728 -1.247112  0.217308   
1 -0.404385  1.582330 -1.320660  1.431040 -0.426050  1.847168  1.067330   
2 -1.635303  0.041872  0.967438  0.716142  0.309850 -1.234978  0.167643   
3 -0.414676  1.581015 -0.990336 -0.345033 -0.878229 -0.373433  1.583075   
4 -0.413356  1.582987 -1.030620 -1.931212 -0.842764 -0.713197  1.405429   

        7         8         9    ...       220       221       222       223  \
0  0.579079  0.788907 -0.304840  ...  0.131358  0.265960 -0.024079 -0.024079   
1 -1.490713 -0.847202  1.004183  ... -0.687092  0.265960 -0.024079 -0.024079   
2  0.673360  0.693979  0.047931  ... -0.277867 -0.771131 -0.024079 -0.024079   
3  0.975936 -1.033335  1.045228  ... -0.277867 -0.252586 -0.024079 -0.024079   
4  0.778604 -1.154322  1.041345  ...  0.131358 -0.667422 -0.024079 -0.024079   

        224       225       226  227  228       229  
0  0.194790 -0.020488 -0.053342  0.0  0.0 -0.020971  
1  1.403979 -0.020488 -0.053342  0.0  0.0 -0.020971  
2 -1.498075 -0.020488 -0.053342  0.0  0.0 -0.020971  
3 -0.772562 -0.020488 -0.053342  0.0  0.0 -0.020971  
4 -1.860832 -0.020488 -0.053342  0.0  0.0 -0.020971  

[5 rows x 230 columns]

In [67]:
X_test=scaler.fit_transform(X_test)
X_test=pd.DataFrame(X_test)
X_test.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


0         1         2         3         4         5         6    \
0 -1.720966  0.036515  0.989078  1.830904  0.906845 -1.284709 -0.102168   
1 -1.720561  0.034530  0.980810  1.830904  0.906845 -1.200860 -0.132972   
2 -1.718940  0.055712  0.980810  1.771581  0.703317 -1.224817  0.009198   
3 -1.718468  0.060346  0.980810  1.771581  0.703317 -1.248774  0.004459   
4 -1.717860  0.045783  0.980810  1.771581  0.703317 -1.236795 -0.061887   

        7         8         9    ...       220       221       222  223  \
0  0.677944  0.914850  0.127925  ... -0.439407 -1.636712 -0.019784  0.0   
1  0.697429  0.926634  0.131266  ...  0.246890 -0.020421 -0.019784  0.0   
2  0.929078  0.677203  0.136279  ...  0.246890 -0.210573 -0.019784  0.0   
3  0.996192  0.580966  0.139064  ... -1.125704  0.264807 -0.019784  0.0   
4  0.881449  0.683095  0.131823  ...  0.246890  1.690946 -0.019784  0.0   

        224  225       226  227  228       229  
0  0.582073  0.0 -0.059444  0.0  0.0 -0.019784  
1 -0.032661  0.0 -0.059444  0.0  0.0 -0.019784  
2  0.459126  0.0 -0.059444  0.0  0.0 -0.019784  
3 -0.524447  0.0 -0.059444  0.0  0.0 -0.019784  
4  0.213233  0.0 -0.059444  0.0  0.0 -0.019784  

[5 rows x 230 columns]

In [68]:
# Load the libraries
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam

In [69]:
length=X_train.shape[1]
num_classes=y_test.shape[1]

def classification_model():
    # create model
    model = Sequential()
    model.add(Dense(length, activation='relu', input_shape=(length,)))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(60, activation= 'relu'))
    model.add(Dense(60, activation = 'relu'))
    model.add(Dense(num_classes, activation='softmax'))
    opt=Adam(lr=0.001)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [73]:
model=classification_model()
model.fit(X_train,y_train,validation_data=(X_test, y_test),epochs=14, verbose=2)
model.evaluate(X_test,y_test,verbose=0)

Train on 50048 samples, validate on 5112 samples
Epoch 1/14
 - 46s - loss: 0.1654 - acc: 0.9436 - val_loss: 0.1007 - val_acc: 0.9673
Epoch 2/14
 - 56s - loss: 0.0682 - acc: 0.9782 - val_loss: 0.0674 - val_acc: 0.9791
Epoch 3/14
 - 44s - loss: 0.0451 - acc: 0.9853 - val_loss: 0.0633 - val_acc: 0.9791
Epoch 4/14
 - 44s - loss: 0.0371 - acc: 0.9885 - val_loss: 0.0701 - val_acc: 0.9787
Epoch 5/14
 - 44s - loss: 0.0302 - acc: 0.9910 - val_loss: 0.0464 - val_acc: 0.9867
Epoch 6/14
 - 43s - loss: 0.0268 - acc: 0.9915 - val_loss: 0.0588 - val_acc: 0.9832
Epoch 7/14
 - 43s - loss: 0.0266 - acc: 0.9915 - val_loss: 0.1240 - val_acc: 0.9714
Epoch 8/14
 - 44s - loss: 0.0246 - acc: 0.9930 - val_loss: 0.0407 - val_acc: 0.9887
Epoch 9/14
 - 44s - loss: 0.0196 - acc: 0.9942 - val_loss: 0.0378 - val_acc: 0.9885
Epoch 10/14
 - 43s - loss: 0.0204 - acc: 0.9938 - val_loss: 0.0389 - val_acc: 0.9890
Epoch 11/14
 - 44s - loss: 0.0149 - acc: 0.9953 - val_loss: 0.0441 - val_acc: 0.9861
Epoch 12/14
 - 44s - loss

[0.042170135279282955, 0.9894366197183099]